### PR TITLE
React router v4

### PR DESCRIPTION
**Pull request by johangirod**
- `react-router` 4.x 🎉 
- `react-resolver` instead of `async-props`
- `react-hot-loader`
- Add a new `HTTPStatus` element, which can be used by the app to specified the
  response status of the server (because of the new `react-router` API)
- Remove react-redux-router, which was not compatible with react-router 4. But in
  the same time, it didn't bring anything besides history. You might want to
  change it for history.push in your action creators. Chances are, that if you were
  using it, you didn't really needed it.
- Refacto, much more cleaner, thanks to the new react-router API. Removing all the
  unecessary koa middlewares

#226 #227 #228

### CODE REVIEW

**Comment by christophehurpeau** [comment]
missing yarn.lock ?

File: package.json [hunk 1]
```diff
@@ -78,11 +76,12 @@
...
     "progress": "^1.1.8",
-    "react-dom": "^15.4.0",
+    "react-dom": "^15.4.2",
     "react-helmet": "^4.0.0",
+    "react-hot-loader": "^3.0.0-beta.6",
     "react-redux": "^5.0.0",
-    "react-router": "^2.8.1",
-    "react-router-redux": "^4.0.5",
+    "react-resolver": "^3.1.0",
+    "react-router-dom": "^4.0.0-beta.5",
```
Comment (christophehurpeau) [inline package.json]: I don't feel confortable using a beta version of react-router if we want to release a 1.0 in vitamin
beta versions can have breaking changes

Comment (alepee) [inline package.json]: I agree on your point @christophehurpeau. Do we have an idea of when it should be released?

File: src/client/index.jsx [hunk 2]
```diff
@@ -42,12 +42,17 @@ function bootstrapClient() {
             }
         });
 
-    if (module.hot && process.env.NODE_ENV !== 'production') {
```
Comment (christophehurpeau) [inline src/client/index.jsx]: TODO: verifier que module.hot est delete en production, ou utiliser le plugin webpack pour le supprimer automatiquement (commen process.env.NODE_ENV)

File: src/shared/store.js [hunk 3]
```diff
@@ -4,19 +4,18 @@ import {
     combineReducers,
     applyMiddleware,
 } from 'redux';
-import { routerReducer, routerMiddleware } from 'react-router-redux';
```
Comment (christophehurpeau) [inline src/shared/store.js]: TODO: extraire la suppression de react-router-redux

**Comment by christophehurpeau** [comment]
passer a react-router 3.0 ?

**Comment by christophehurpeau** [comment]
remove js stringify state to have better perf ?

**Comment by christophehurpeau** [comment]
https://medium.com/node-security/the-most-common-xss-vulnerability-in-react-js-applications-2bdffbcc1fa0#.tm3hd6riw

**Comment by johangirod** [comment]
Closing this until react-router is released

**Pull request closed without merging**